**Installation of NumPy**

In [ ]:
pip install numpy

**Checking NumPy Version**

In [ ]:
import numpy as np
print(np.__version__)

1.25.2


**Create a NumPy ndarray Object**

In [ ]:
import numpy as np
l =[1,2,3,4,5]
print(l)
arr = np.array([1, 2, 3, 4, 5])
print(arr)
print(type(arr))

[1, 2, 3, 4, 5]
[1 2 3 4 5]
<class 'numpy.ndarray'>


# **Differnce between List and traditional Array in python**

**1.Data Type:**

* The array module provides a fixed-type array implementation, meaning all elements must be of the same data type.

* Lists, on the other hand, can contain elements of different data types.

**2.Mutability:**

* Arrays created with the array module are mutable, meaning you can change the values of individual elements after creation.

* Lists are also mutable, allowing you to change, add, or remove elements after creation.

**3.Type Flexibility:**

* The array module requires you to specify the data type of elements when creating an array.
* This enforces homogeneity within the array.
* Lists in Python are more flexible, allowing elements of different data types within the same list.

**4.Memory Efficiency:**

* Arrays from the array module can be more memory efficient compared to lists because they store elements of a fixed data type.
* Lists may require more memory because they can store elements of different data types, leading to potential memory overhead.










In [ ]:
# Using the array module:
import array

# Creating an array of integers
int_array = array.array('i', [1, 2, 3, 4, 5])

# Modifying an element
int_array[2] = 10

# Printing the array
print("Array:", int_array)


SyntaxError: invalid syntax (<ipython-input-9-cade6462db9b>, line 5)

In [ ]:
# Using Python Lists:
# Creating a list with elements of different data types
my_list = [1, 2.5, 'hello']

# Modifying an element
my_list[1] = 5.0

# Printing the list
print("List:", my_list)


List: [1, 5.0, 'hello']


* In both cases, we create a data structure and modify one of its elements.
* The key difference lies in the type flexibility and memory usage. The array module enforces a fixed data type for all elements, while lists allow for heterogeneity.
* Depending on the use case, you might choose one over the other for memory efficiency, type requirements, or specific functionality needs.

# Difference between Array and NumPy Array
**Normal array Module:**
* The array module in Python provides a way to create arrays with a fixed type of elements.
* It is similar to lists but has a fixed type for all elements.
It is less flexible compared to NumPy arrays and lists in terms of operations and functionality.

In [ ]:
import array

# Creating an array of integers
int_array = array.array('i', [1, 2, 3, 4, 5])

# Accessing elements
print("Array:", int_array)
print("First Element:", int_array[0])

# Modifying elements
int_array[2] = 10
print("Modified Array:", int_array)

Array: array('i', [1, 2, 3, 4, 5])
First Element: 1
Modified Array: array('i', [1, 2, 10, 4, 5])


# NumPy Arrays:
* NumPy is a powerful library in Python for numerical computations.
* NumPy arrays are homogeneous, meaning they can only contain elements of the same data type.
* NumPy arrays offer extensive functionality for mathematical operations and are widely used in scientific computing and data analysis.

In [ ]:
import numpy as np

# Creating a NumPy array of integers
int_array_np = np.array([1, 2, 3, 4, 5])

# Accessing elements
print("NumPy Array:", int_array_np)
print("First Element:", int_array_np[0])

# Modifying elements
int_array_np[2] = 10
print("Modified NumPy Array:", int_array_np)


NumPy Array: [1 2 3 4 5]
First Element: 1
Modified NumPy Array: [ 1  2 10  4  5]


**In both examples:**

* We create arrays (int_array with the array module and int_array_np with NumPy) and access elements in the same way.
* However, NumPy arrays offer additional functionality, such as broadcasting and vectorized operations, making them more suitable for numerical computations.
* NumPy arrays also have better memory efficiency and performance compared to the normal array module.

In [ ]:
import array

# Creating an array of integers
int_array1 = array.array('i', [1, 2])
int_array2 = array.array('i', [3, 4])
mul_array = int_array1 * int_array2
mul_array

print("Modified Array:", mul_array)

TypeError: can't multiply sequence by non-int of type 'array.array'

**Vectorized Operations:**

* Vectorized operations allow NumPy to perform element-wise operations on arrays without the need for explicit looping.
* These operations are optimized and much faster than using traditional loop-based operations.

In [ ]:
import numpy as np

# Creating an array of integers
int_array1 = np.array([1, 2])
int_array2 = np.array([3, 4])
mul_array = int_array1 * int_array2
mul_array

print("Result:", mul_array)

Result: [4 6]


# Difference between NumpyArray and List

In [ ]:
list_obj = [1,2,3]
print(list_obj)


[1, 2, 3]


In [ ]:
import numpy as np
array = np.array(list_obj)
type(array)

numpy.ndarray

In [ ]:
list_obj = [1,2,'pc']
list_obj

[1, 2, 'pc']

In [ ]:
#array = np.array([1,2,'pc'], dtype = 'int')
array = np.array([1,2,'pc'])
array

array(['1', '2', 'pc'], dtype='<U21')

# Effect of operations on Numpy array and Python Lists

In [ ]:
list1 = [10,20,30]
list2 = [10,20,30]
print(list1 + list2)

[10, 20, 30, 10, 20, 30]


In [ ]:
array1 = np.array([1,2,3])
array2 = np.array([1,2,3])
array1 + array2

array([2, 4, 6])

In [ ]:
z = [1,2,3]


In [ ]:
list1 * z

NameError: name 'list1' is not defined

In [ ]:
array1 * z

array([1, 4, 9])

# Memory consumption between range and lists object

In [ ]:
import sys

# Using range to generate a sequence of numbers
my_range = range(1000000)  # Create a range object with a million numbers

# Check the memory usage of the range object
print("Memory used by range object:", sys.getsizeof(my_range), "bytes")

Memory used by range object: 48 bytes


In [ ]:
# Converting range object to a list
my_list = list(my_range)  # Convert the range object to a list

# Check the memory usage of the list
print("Memory used by list object:", sys.getsizeof(my_list), "bytes")

Memory used by list object: 8000056 bytes


#Memory Usage Difference
**Range Object:**

**Memory used:** Minimal (approximately 48 bytes).

**Reason:** Stores only start, stop, and step values without generating all numbers in memory.

**List Object:**

**Memory used:** Substantial (approximately 8,000,056 bytes for 1,000,000 integers).

**Reason:** Stores each number as a separate integer in memory, resulting in higher memory consumption.

**Conclusion**
**Efficiency:** Use range when you need to iterate over a sequence of numbers without storing them all in memory. This is useful for large sequences.

**Memory Usage:** Be cautious when converting a range to a list as it results in significant memory usage, especially for large sequences.

* This explanation highlights the efficiency of the range object and the higher memory consumption of the list object, providing insight into their usage based on memory constraints.

# Memory consumption between Numpy array and lists

In [ ]:
import numpy as np
import sys

# Create a Python list containing integers from 0 to 999
py_list = range(1000)
#py_list = [i for i in range(1000)]

# Create a NumPy array containing integers from 0 to 999
np_array = np.arange(1000)

# Check the memory usage of the Python list
print("Memory used by Python range:", sys.getsizeof(py_list), "bytes")

# Check the memory usage of the NumPy array
print("Memory used by NumPy array:", np_array.nbytes, "bytes")


Memory used by Python range: 48 bytes
Memory used by NumPy array: 8000 bytes


**Python List:**

**Memory Consumption:**

* The memory used by the Python list includes the overhead for storing Python objects.
* Each element in the list is a full-fledged Python object, which includes additional metadata.

**Output:**
* The sys.getsizeof(py_list) function returns the memory size of the list object itself, but it does not include the memory used by the actual elements in the list.
* To get the total memory usage, you would need to sum the sizes of the individual elements.

**NumPy Array:**

**Memory Consumption:**
* The NumPy array stores data in a contiguous block of memory with a fixed data type, which reduces the overhead and results in more efficient memory usage.

**Output:**

* The np_array.nbytes attribute returns the total memory size of the array, which includes all elements in the array.
* This gives an accurate representation of the memory used by the NumPy array.


**Python List:**

* Flexible but less memory-efficient due to the overhead of storing each element as a separate Python object.

**NumPy Array:**

* More memory-efficient for numerical data, storing all elements in a contiguous block with a fixed data type.

* The difference in memory usage highlights the advantage of NumPy arrays for large numerical datasets, where memory efficiency is crucial.

# Time comparison between Numpy array and Python lists

In [ ]:
#using Lists
import time

# Start time
start_time = time.time()

# Using Python list comprehension to generate a list of numbers from 1 to 1,000,000
python_list = [i for i in range(1, 1000001)]

# Calculate the sum of squares
sum_of_squares = sum([x**2 for x in python_list])

# End time
end_time = time.time()

# Print the result and the time taken
print("Sum of squares using Python list:", sum_of_squares)
print("Time taken with Python list:", end_time - start_time, "seconds")


Sum of squares using Python list: 333333833333500000
Time taken with Python list: 0.40378594398498535 seconds


In [ ]:
#Using NumPy Arrays:

import numpy as np
import time

# Start time
start_time = time.time()

# Using NumPy array to generate numbers from 1 to 1,000,000
numpy_array = np.arange(1, 1000001)

# Calculate the sum of squares using NumPy's vectorized operations
sum_of_squares = np.sum(numpy_array**2)

# End time
end_time = time.time()

# Print the result and the time taken
print("Sum of squares using NumPy array:", sum_of_squares)
print("Time taken with NumPy array:", end_time - start_time, "seconds")


Sum of squares using NumPy array: 333333833333500000
Time taken with NumPy array: 0.007634639739990234 seconds


# Explanation
* In both code snippets, we calculate the sum of squares of numbers from 1 to 1,000,000.
* With Python lists, we use list comprehension to generate the list of numbers and then sum the squares using the sum function.
* With NumPy arrays, we use np.arange to generate the array of numbers and then use NumPy's vectorized operations (**2 and np.sum) to calculate the sum of squares.
* The output shows that both methods yield the same result, but the time taken with NumPy arrays is significantly shorter compared to Python lists.
* This demonstrates the efficiency of NumPy arrays for numerical computations.

#Advantages of using Numpy Arrays Over Python Lists:

* Consumes less memory.
* Fast as compared to the python List.
* Convenient to use.

# NDim, Shape, ReShape

**ndim**
* Level of depth of array

#Shape
* In NumPy, the shape attribute is used to get the dimensions of an array.
* This attribute returns a tuple representing the dimensions of the array.
* You don't need to call it as a function; simply access it as an attribute.
* Here’s how to use the shape attribute for zero-dimensional, one-dimensional, two-dimensional, and three-dimensional NumPy arrays:

**Zero-Dimensional Array (Scalar)**
* A zero-dimensional array is essentially a scalar value.

**One-Dimensional Array (Vector)**
* A one-dimensional array is a simple list of elements.

**Two-Dimensional Array (Matrix)**
* A two-dimensional array is like a matrix with rows and columns.

**Three-Dimensional Array (Tensor)**
* A three-dimensional array can be thought of as a stack of matrices.

In [ ]:
zeroDArray = np.array(23)
zeroDArray

array(23)

In [ ]:
type(zeroDArray)

numpy.ndarray

In [ ]:
zeroDArray.ndim
#np.ndim(zeroDArray)

0

**Zero-Dimensional Array:**

Shape is (), indicating a single scalar.

In [ ]:
zeroDArray.shape

()

In [ ]:
oneDArray = np.array([1, 2, 3, 4, 5])
oneDArray

array([1, 2, 3, 4, 5])

In [ ]:
oneDArray.ndim
#np.ndim(oneDArray)

1

**One-Dimensional Array:**

Shape is (n,), where n is the number of elements.

In [ ]:
#shape() --> return tuple specifying no of elements
oneDArray.shape

(5,)

In [ ]:
twoDArray = np.array([[1, 2, 3], [4, 5, 6]])
twoDArray

array([[1, 2, 3],
       [4, 5, 6]])

In [ ]:
twoDArray.ndim
#np.ndim(twoDArray)

2

**Two-Dimensional Array:**

Shape is (rows, columns).


In [ ]:
#shape() --> return tuple specifying indices and no of elements
twoDArray.shape

(3, 3)

In [ ]:
threeDArray = np.array([[[1, 2, 3], [4, 5, 6]], [[7, 8, 9], [10, 11, 12]]])
threeDArray

array([[[ 1,  2,  3],
        [ 4,  5,  6]],

       [[ 7,  8,  9],
        [10, 11, 12]]])

In [ ]:
threeDArray.ndim
#np.nidm(threeDArray)

3

**Three-Dimensional Array:**

Shape is (depth, rows, columns)

In [ ]:
threeDArray.shape

(2, 2, 3)

**Reshape**
* The reshape method in NumPy is used to change the shape of an array without changing its data.
* This method allows you to create a new view of the array with the desired shape.
* The total number of elements in the new shape must be the same as the original array.

In [ ]:
#Reshaping a 1D Array to a 2D Array
import numpy as np

# Creating a 1D array
original_array = np.array([1, 2, 3, 4, 5, 6])

# Reshaping the 1D array to a 2D array (2 rows, 3 columns)
reshaped_array = original_array.reshape((2, 3))

print("Original array:")
print(original_array)
print("Reshaped array:")
print(reshaped_array)

Original array:
[1 2 3 4 5 6]
Reshaped array:
[[1 2 3]
 [4 5 6]]


In [ ]:
#reshape synatx: reshape(array_name,new_dimensions)

print("TwoDArray:",twoDArray)
print("Before reshaping:",np.shape(twoDArray))
newArray = twoDArray.reshape(3,2)
#newArray = np.reshape(newArray,(3,2))
print("After reshaping:",np.shape(newArray))
print(newArray)

TwoDArray: [[1 2 3]
 [4 5 6]]
Before reshaping: (2, 3)
After reshaping: (3, 2)
[[1 2]
 [3 4]
 [5 6]]


In [ ]:
print("ThreeDArray:",threeDArray)
print("Before reshaping:",np.shape(threeDArray))


ThreeDArray: [[[ 1  2  3]
  [ 4  5  6]]

 [[ 7  8  9]
  [10 11 12]]]
Before reshaping: (2, 2, 3)


In [ ]:
#newArray = threeDArray.reshape(2,3)
tempArray = threeDArray.reshape(1,1,12)
print("After reshaping:",np.shape(tempArray))
print(tempArray)

After reshaping: (1, 1, 12)
[[[ 1  2  3  4  5  6  7  8  9 10 11 12]]]


In [ ]:
import numpy as np

# Create a 3D NumPy array
array_3d = np.array([[[1, 2], [3, 4],[5, 6], [7, 8], [9, 10], [11, 12]]])

# Get the shape of the 3D array
shape_3d = array_3d.shape

print("Shape of the 3D array:", shape_3d)


Shape of the 3D array: (1, 6, 2)


# Accessing of an Array

In [ ]:
oneDArray

array([1, 2, 3, 4, 5])

In [ ]:
oneDArray[3]

1

In [ ]:
oneDArray[-3]

In [ ]:
twoDArray

array([[1, 2, 3],
       [4, 5, 6]])

In [ ]:
twoDArray[1,0]

4

In [ ]:
twoDArray[0][1]

2

In [ ]:
threeDnewArray = np.array( [ [[10,20,30], [40,50,60]], [[70,80,90], [100,110,120]]  ])

In [ ]:
threeDnewArray.shape

(2, 2, 3)

In [ ]:
threeDnewArray[1,1,2]

60

# Slicing of an Array

In [ ]:
oneDArray

array([500,   2,   3,   4,   5])

In [ ]:
oneDArray[ : :]

array([1, 2, 3, 4, 5])

In [ ]:
oneDArray[ 2: 5:]

array([3, 4, 5])

In [ ]:
oneDArray[ : -3:]

array([1, 2])

In [ ]:
oneDArray[ 2: 4: 2]

array([3])

In [ ]:
twoDArray

array([[1, 2, 3],
       [4, 5, 6]])

In [ ]:
twoDArray[0,1:3]

array([4, 5, 6])

In [ ]:
twoDArray[2, 1:]

In [ ]:
twoDArray

array([[1, 2, 3],
       [4, 5, 6]])

In [ ]:
twoDArray[0:2,1:3]

array([[2, 3],
       [5, 6]])

In [ ]:
twoDArray

array([[1, 2, 3],
       [4, 5, 6]])

In [ ]:
twoDArray[0:2,1:2]

array([[2],
       [5]])

In [ ]:
twoDArray[1:2,1:3]

array([[5, 6]])

In [ ]:
threeDArray

In [ ]:
threeDArray[0,1,2]

In [ ]:
threeDArray[0,0,:2]

# Axis
# 0 --> Column
# 1 --> Row

# Sort

In [ ]:
sortArray = [[40, 10, 30], [10, 5, 15]]
sortArray

[[40, 10, 30], [10, 5, 15]]

In [ ]:
np.sort(sortArray)

array([[10, 30, 40],
       [ 5, 10, 15]])

In [ ]:
np.sort(sortArray,axis=0)

array([[10,  5, 15],
       [40, 10, 30]])

# Copy vs View
# copy() --> External copy and creates two arrays
#            If any one element is modified then it will not
#            reflect to another array             
# view() --> Cretae a reference to existing array and creates only one array
#            If any one element is modified then it will be
#            reflect on original array

In [ ]:
copyArray = np.copy(oneDArray)
copyArray[2] = 50
print("Original array",oneDArray)
print("Copied Array",copyArray)
print("Checking oneDArray", oneDArray)

Original array [1 2 3 4 5]
Copied Array [ 1  2 50  4  5]
Checking oneDArray [1 2 3 4 5]


In [ ]:
viewArray = oneDArray.view()
viewArray[0]=500
print("Original array",oneDArray)
print("view array",viewArray)

Original array [500   2   3   4   5]
view array [500   2   3   4   5]


# Datatypes

In [ ]:
# Data Types
# i --> int f --> float s ---> string u ---> unicode strings M --> datetime o ---> object
#oneDArray.dtype
x = np.array([10,20,30],dtype='U')
x

array(['10', '20', '30'], dtype='<U2')

# Astype

In [ ]:
y = np.array(["1", "2", "3"]).astype('i')
y

array([1, 2, 3], dtype=int32)

# Searching (Where)

* returns index of given value




In [ ]:
oneDArray

array([500,   2,   3,   4,   5])

In [ ]:
np.where(oneDArray==4)

(array([3]),)


# searchsorted()

* Binary Search
* Return the index where the specified value should be inserted
* Elements should be in sorted order







In [ ]:
z = np.array([10,20,30,40])
np.searchsorted(z, [5, 15, 45 ,50])

array([0, 1, 4, 4])

# Splitting
# Syntax: split(array, no of splits)
# Note :                       No of Splits--> Equal Division

In [ ]:
oneDArray_new = np.array([1, 2, 3,   4,   5, 6])

In [ ]:
np.split(oneDArray_new,2)

[array([1, 2, 3]), array([4, 5, 6])]

In [ ]:
np.split(oneDArray_new,4)

ValueError: array split does not result in an equal division

# array_split()

# Syntax: array_split(array, no of splits)
# Note :                       No of Splits--> Equal Division is not required

In [ ]:
np.array_split(oneDArray_new,4)

[array([1, 2]), array([3, 4]), array([5]), array([6])]

# hsplit()
equal division horizontal manner similar to split()

In [ ]:
np.hsplit(oneDArray_new,2)
#np.hsplit(oneDArray_new,4)

ValueError: array split does not result in an equal division

# vsplit()
not applicable for 1 D Array
Minimum of 2D array is required to implement vsplit()
equal divison vertical manner

In [ ]:
#np.vsplit(oneDArray,2)
np.vsplit(twoDArray,2)


[array([[1, 2, 3]]), array([[4, 5, 6]])]

# Linspace

In [ ]:
#linspace(start,stop,number=50,endpoint=true,
#          retstep=false, dtype=float)
np.linspace(0,5, num =  15, retstep = True)

(array([0. , 2.5, 5. ]), 2.5)

# Zeros, Ones, Full, Eye

In [ ]:
np.zeros((2,2))
#np.zeros((2,2),dtype= 'i')

array([[0., 0.],
       [0., 0.]])

In [ ]:
np.ones((2,2))

array([[1., 1.],
       [1., 1.]])

In [ ]:
np.full((2,2), 20)

array([[20, 20],
       [20, 20]])

In [ ]:
np.eye(3)
np.eye(5)

#temp = np.eye(3)
#temp
#temp.reshape(3)

array([[1., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1.]])

In [ ]:
np.full((3,3),29)

array([[29, 29, 29],
       [29, 29, 29],
       [29, 29, 29]])

In [ ]:
import numpy as np
ab, ac = (4, 2)
b = np.linspace(0, 1, ab)
c = np.linspace(0, 1, ac)
bd, cd = np.meshgrid(b, c)
print(b)
print(c)
print(bd)
print(cd)

# Arrays with existing Data, Astype

In [ ]:
#Arrays with existing Data
#converting one type to another type
#C --> ROW & F --> COL
a = np.array([10,20,30])
res = np.asarray(a, dtype = float, order = 'F')
print(res)

In [ ]:
b = np.array([[10,20,30],[40, 50 ,60]])
print(b)

In [ ]:
res = np.asarray(b, dtype = float, order = 'F')
print(res)

In [ ]:
#NDITER ---> Single dimension
for i in np.nditer(res):
    print(i)

In [ ]:
a = np.array([10,20,30])
#print(a)
print(a.astype(float))

# Joining

# Concatenate()
--- Join two arrys with respect to given axis(with respect to rows & cols)

--- concatenate((a,b) ,  axis)
by default axis is 0

a&b ---> names of arrays

axis ---> 0(columns) & 1(rows)







In [ ]:
#concatenate
a = np.array([[10,20,30],[40,50,60]])
b = np.array([[70,80,90],[100,110,120]])

In [ ]:
a

array([[10, 20, 30],
       [40, 50, 60]])

In [ ]:
b

array([[ 70,  80,  90],
       [100, 110, 120]])

In [ ]:
np.concatenate((a,b))
#by default axis is 0(column)

array([[ 10,  20,  30],
       [ 40,  50,  60],
       [ 70,  80,  90],
       [100, 110, 120]])

In [ ]:
np.concatenate((a,b),axis = 1)

array([[ 10,  20,  30,  70,  80,  90],
       [ 40,  50,  60, 100, 110, 120]])

#Stack

In [ ]:
a

array([[10, 20, 30],
       [40, 50, 60]])

In [ ]:
b

array([[ 70,  80,  90],
       [100, 110, 120]])

In [ ]:
#concatenation is single dimensional
#stack is  two dimensional
np.stack((a,b))

array([[[ 10,  20,  30],
        [ 40,  50,  60]],

       [[ 70,  80,  90],
        [100, 110, 120]]])

In [ ]:
np.stack((a,b), axis = 1)

array([[[ 10,  20,  30],
        [ 70,  80,  90]],

       [[ 40,  50,  60],
        [100, 110, 120]]])

#hStack
Similar to concatenate axis = 1 (row)

In [ ]:
a

In [ ]:
b

In [ ]:
np.hstack((a,b))

array([[ 10,  20,  30,  70,  80,  90],
       [ 40,  50,  60, 100, 110, 120]])

# VStack
Similar to concatenate axis = 0 (Column)

In [ ]:
a

In [ ]:
b

In [ ]:
np.vstack((a,b))

array([[ 10,  20,  30],
       [ 40,  50,  60],
       [ 70,  80,  90],
       [100, 110, 120]])

# DStack
based on the depth of the array elements will join

In [ ]:
a

array([[10, 20, 30],
       [40, 50, 60]])

In [ ]:
b

array([[ 70,  80,  90],
       [100, 110, 120]])

In [ ]:
np.dstack((a,b))

array([[[ 10,  70],
        [ 20,  80],
        [ 30,  90]],

       [[ 40, 100],
        [ 50, 110],
        [ 60, 120]]])

# Matrix Opeartions

In [ ]:
a = np.array([[1, 2, 3],
              [4, 5, 6]])
b = np.array([[1, 1, 1],
              [1, 1, 1]])
print(np.add(a,b))

[[2 3 4]
 [5 6 7]]


In [ ]:
a = np.array([[1, 2, 3],
              [4, 5, 6]])
b = np.array([[1, 1, 1],
              [1, 1, 1]])
np.subtract(a,b)

array([[0, 1, 2],
       [3, 4, 5]])

In [ ]:
a = np.array([[1, 2, 3],
              [4, 5, 6],
              [7,8,9]])
b = np.array([[1, 2, 3],
              [1, 1, 1],
              [1,1,1]])
#np.multiply(a,b)
np.dot(a,b)
#a @ b

array([[ 6,  7,  8],
       [15, 19, 23],
       [24, 31, 38]])